shuttle/src/runtime/execution.rs

```diff
@@ -89,159 +151,176 @@ impl Execution {
                 caller,
             );
 
-            // Run the test to completion
-            while self.step(config) {}
+                // Run the test to completion
+                match self.run_to_competion(config.immediately_return_on_panic) {
+                    Ok(()) => {},
+                    Err(e) => {
+                        match e {
+                            StepError::TaskFailure(payload) => {
+                                let task_name = ExecutionState::failing_task();
+                                persist_failure(config);
+                                eprintln!("test panicked in task '{task_name}'");
+
+                                panic::resume_unwind(payload);
+                            }
+                            StepError::Deadlock => {
+                                let blocked_tasks = ExecutionState::with(|state|
+                                    state
+                                    .tasks
+                                    .iter()
+                                    .filter(|t| !t.finished())
+                                    .map(|t| {
+                                        format!(
+                                            "{} (task {:?}{}{}){}",
+                                            t.name().unwrap_or_else(|| "<unknown>".to_string()),
+                                            t.id(),
+                                            if t.detached { ", detached" } else { "" },
```

Comment:
Instead of inlining all this code here, could just define a `format_for_deadlock`